src/ray/object_manager/object_manager_client_pool.h

```diff
@@ -0,0 +1,101 @@
+// Copyright 2020 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include "absl/base/thread_annotations.h"
+#include "absl/container/flat_hash_map.h"
+#include "absl/strings/str_cat.h"
+#include "absl/synchronization/mutex.h"
+#include "ray/common/id.h"
+#include "ray/gcs/gcs_client.h"
+#include "ray/rpc/object_manager/object_manager_client.h"
+
+using absl::optional;
+using std::shared_ptr;
+
+namespace ray {
+
+using ObjectManagerClientFactoryFn =
+    std::function<std::shared_ptr<rpc::ObjectManagerClient>(const rpc::Address &)>;
+
+class IObjectManagerClientPool {
+ public:
+  virtual ~IObjectManagerClientPool() {}
+
+  /// Return an existing ObjectManagerClient if exists, and connect to one if it does
+  /// not. The returned pointer is borrowed, and expected to be used briefly.
+  virtual optional<shared_ptr<rpc::ObjectManagerClient>> GetOrConnectByID(
```

Comment:
> The returned pointer is borrowed, and expected to be used briefly.

this looks overkill? at least this is inconsistent with the API. shared_ptr means shared ownership so it's not borrowed but rather hold a shared ownership.

Reply:
Is the `absl::optional<>` wrapper necessary to represent an error condition, or a non-empty `absl::optional<>` always contains a non-empty `std::shared_ptr` here?